### DISCUSSION TITLE
SetState dim value range?

### DISCUSSION THREAD

**Discussion by Miq1** [vintlabs/fauxmoESP#139] [General]
I seem not to have understood yet the ``value``  parameter for the SetState callback. 

I tried to catch a certain percentage by ``if (value >= 40 && value <= 49)``, but it seems never to get these values when called from Alexa. 

Unfortunately I cannot print the value received, since the device sits in a power plug with just two LEDs and no Serial available.

Is someone able to enlighten me?

**Comment by Miq1** [comment]
Addendum: I used the LEDs to blink the high and low nibbles of value, and both seem to be 0 - also value must be 0 all the time. 

Do I need to tell Alexa or the Fauxmo lib that it is a dimmable device?

**Comment by pvint** [comment]
Just did a quick test with the following in my sketch:
```C
    fauxmo.addDevice(DEVICE_NAME);  

    fauxmo.onSetState([](unsigned char device_id, const char * device_name, bool state, unsigned char value) {
      Serial.printf("[MAIN] Device #%d (%s) state: %s value: %d\n", device_id, device_name, state ? "ON" : "OFF", value);
    });
```

Gave Alexa the following commands, and my serial output follows:

"Set light to zero":
```
[MAIN] Device #0 (Zulu) state: ON value: 1
[MAIN] Device #0 (Zulu) state: OFF value: 1
```

"Set light to 100":
```
[MAIN] Device #0 (Zulu) state: ON value: 1
[MAIN] Device #0 (Zulu) state: ON value: 254
```

"Dim light":
```
[MAIN] Device #0 (Zulu) state: ON value: 191
```

"Set light to 50":
```
[MAIN] Device #0 (Zulu) state: ON value: 128
```


*Edit: Note that `value` is an 8 bit value, not a percentage

**Comment by Miq1** [comment]
Well, thank you for the quick reply, but no, does not work for me.

I let the Alexa app detect the device as Philips Hue device, no Bluetooth enabled and a V1 hub. After the search the device is displayed in the app as a small house icon with a WiFi symbol on it - no light bulb icon. 

If I tell Alexa to dim the device, it is switche on regardless and value obviously is zero. 

Is there another recommended procedure?

**Comment by pvint** [comment]
>I let the Alexa app detect the device as Philips Hue device, no Bluetooth enabled and a V1 hub. After the search the device is displayed in the app as a small house icon with a WiFi symbol on it - no light bulb icon.

This is normal, sadly.

Perhaps try having it flash the LED `value` times, on for 500ms and off for 500ms? If you can't get some info from that, the only other thing I can suggest is to either get a serial connection or find some other way of logging what's happening.
I've often used https://github.com/arcao/Syslog to have an ESP log to syslog on a Linux machine.

**Reply by Miq1** [reply]
That is exactly what I did:
```
void SetState(uint8_t device_id, const char * device_name, bool state, uint8_t value) {
  uint8_t high = (value >> 4) & 0x0F;
  uint8_t low = value & 0x0F;

  while (high--) {
    digitalWrite(POWER_LED, LOW); delay(500);
    digitalWrite(POWER_LED, HIGH); delay(500);
  }
  while (low--) {
    digitalWrite(SIGNAL_LED, LOW); delay(500);
    digitalWrite(SIGNAL_LED, HIGH); delay(500);
  }
  // Special command for "find me" mode?
  if (value >=40 && value <= 49) {
    // Yes, start find mode
    SignalLed.start(100, FINDBLINK);
  } else {
    // No, switch mode
    SignalLed.stop();
    if (state) { // ON
      Testschalter = true;
      digitalWrite(POWER_LED, LOW);
      digitalWrite(RELAY, HIGH);
    } else { // OFF
      Testschalter = false;
      digitalWrite(POWER_LED, HIGH);
      digitalWrite(RELAY, LOW);
    }
  }
}
```
Since Fauxmo is already running on port 80: can I set up another web server to display the values in parallel? There is not a single GPIO available any more on the ESP8285 in that device.

**Comment by pvint** [comment]
You might be getting into trouble because it's spending too long in the callback function. Rather than doing the `digitalWrite(); delay()` in the callback, try setting a flag and doing that in the main loop.

**Reply by Miq1** [reply]
Sure. But at the moment it immediately switches the relay, so neither the high nor the low loop is ever entered. I will try to set up a telnet server to log some debug output. Tomorrow's another day. 😉

**Comment by Miq1** [comment]
Found it by logging to a telnet session.... The problem partly was in front of the keyboard, as ever. I did not realize that Alexa keeps sending the same value when switching the device on and off, so my program always reacted as if the dim value had been set anew.

Looks like my idea to use different dim values to trigger different reactions was not that bright; it relies on new dim values for on and off commands, that simply are not given. I now am disregarding the value if the iswitch state has changed, which mostly is solving the issue for me.

Thank you for your thoughts.
